lib/trento/domain/sap_system/sap_system.ex

```diff
@@ -840,6 +840,7 @@ defmodule Trento.Domain.SapSystem do
          %SapSystem{
            sap_system_id: sap_system_id,
            database: %Database{
+             deregistered_at: nil,
```

Comment:
I think we can have some unsupported scenario with this clause. With this conditions, we only check for `deregistered_at` when the list of `instances` is empty. 
What if we have:
- Scale-out deployment with 2 primary and 2 secondary
- Already deregistered database (all primaries were deregistered)
- 1 of the secondaries is deregistered now
At this point the function would go for the next function, emitting the `DatabaseDeregistered` event again, when it shouldn't.

I think we should do the same with do above in `maybe_emit_sap_system_deregistered_event` and `when not is_nil(database_deregistered_at) do`.

Or you add this same clause in the next `maybe_emit_database_deregistered_event` function. I would bet for the 1st option, so we don't need this `when` twice

What do you think?

Reply:
I will write a test for it so we track this scenario, thanks for pointing me out this thing!

I would go for the first scenario, fixing it asap :D

Reply:
Add a test for this scenario, and added the clause in the guard, we not emit the database deregistered event if it's already registered, IN ANY CASE, with or without instances.

The test passes, and event the new test for this scenario pass (was ofcourse not passing before)